p2p/ipld/read.go

```diff
@@ -3,16 +3,223 @@ package ipld
 import (
 	"context"
 	"errors"
+	"fmt"
 	"math"
+	"math/rand"
 
 	"github.com/ipfs/go-cid"
 	coreiface "github.com/ipfs/interface-go-ipfs-core"
 	"github.com/ipfs/interface-go-ipfs-core/path"
+	"github.com/lazyledger/lazyledger-core/p2p/ipld/plugin/nodes"
+	"github.com/lazyledger/lazyledger-core/types"
+	"github.com/lazyledger/rsmt2d"
 )
 
-// /////////////////////////////////////
+// //////////////////////////////////////
+//	Retrieve Block Data
+// ////////////////////////////////////
+
+const baseErrorMsg = "failure to retrieve block data:"
+
+var ErrEncounteredTooManyErrors = fmt.Errorf("%s %s", baseErrorMsg, "encountered too many errors")
+var ErrTimeout = fmt.Errorf("%s %s", baseErrorMsg, "timeout")
+
+// RetrieveBlockData asynchronously fetches block data using the minimum number
+// of requests to IPFS. It fails if one of the random samples sampled is not available.
+func RetrieveBlockData(
+	ctx context.Context,
+	dah *types.DataAvailabilityHeader,
+	api coreiface.CoreAPI,
+	codec rsmt2d.CodecType,
+) (types.Data, error) {
+	edsWidth := len(dah.RowsRoots)
+	sc := newshareCounter(ctx, uint32(edsWidth))
+
+	// convert the row and col roots into Cids
+	rowRoots := dah.RowsRoots.Bytes()
+	colRoots := dah.ColumnRoots.Bytes()
+
+	// sample 1/4 of the total extended square by sampling half of the leaves in
+	// half of the rows
+	for _, row := range uniqueRandNumbers(edsWidth/2, edsWidth) {
+		for _, col := range uniqueRandNumbers(edsWidth/2, edsWidth) {
+			rootCid, err := nodes.CidFromNamespacedSha256(rowRoots[row])
+			if err != nil {
+				return types.Data{}, err
+			}
+
+			go sc.retrieveShare(rootCid, true, row, col, api)
+		}
+	}
```

Comment:
updated to only sampling 1/4 of the square, which reduced the number of goroutines by 8x. If we ever need to make an "`ImpatientRetrieveBlockData`" we only have to modify this section.